cupy/cuda/stream.pyx

```diff
@@ -24,42 +30,53 @@ cdef class _ThreadLocal:
             tls = _thread_local.tls = _ThreadLocal()
         return <_ThreadLocal>tls
 
+    cdef void push_stream(self, stream, int device_id) except*:
+        if self.prev_stream_ref_stack[device_id] is None:
+            self.prev_stream_ref_stack[device_id] = []
+        prev_stream_ref = self.get_current_stream_ref(device_id)
+        self.prev_stream_ref_stack[device_id].append(prev_stream_ref)
+        # record device_id to prevent from popping the wrong stream at exit
+        self.current_device_id_stack.append(device_id)
+        self.set_current_stream(stream)
+
+    cdef void pop_stream(self) except*:
+        cdef int device_id = self.current_device_id_stack.pop()
+        prev_stream_ref = self.prev_stream_ref_stack[device_id].pop()
+        self.set_current_stream_ref(prev_stream_ref)
+
     cdef set_current_stream(self, stream):
         cdef intptr_t ptr = <intptr_t>stream.ptr
-        stream_module.set_current_stream_ptr(ptr)
-        self.current_stream = <intptr_t>ptr
-        self.current_stream_ref = weakref.ref(stream)
+        cdef int device_id = stream.device_id
+        if device_id == -1:
```

Comment:
Yes I am looking into this , but i doubt it can be removed. For `null` and `ptds` singletons we still need to keep track the device IDs, as they do not carry the current device info (set to -1). But maybe I am just overlooking something.

Reply:
Thank you @leofang, I agree that it's needed for null/ptds.